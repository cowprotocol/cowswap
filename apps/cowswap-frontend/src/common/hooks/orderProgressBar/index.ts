export * from './useOrderProgressBarProps'
export * from './types'
export * from './ProgressBarV2ExecutingOrdersUpdater'
