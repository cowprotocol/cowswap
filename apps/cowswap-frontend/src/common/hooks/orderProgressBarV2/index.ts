export * from './useOrderProgressBarV2Props'
export * from './types'
export * from './ProgressBarV2ExecutingOrdersUpdater'
