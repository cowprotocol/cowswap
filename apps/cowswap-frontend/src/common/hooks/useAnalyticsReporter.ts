import { useEffect } from 'react'

import {
  Dimensions,
  GOOGLE_ANALYTICS_CLIENT_ID_STORAGE_KEY,
  googleAnalytics,
  PixelEvent,
  sendAllPixels,
} from '@cowprotocol/analytics'
import { usePrevious } from '@cowprotocol/common-hooks'
import {
  getConnectionName,
  getIsMetaMask,
  getWeb3ReactConnection,
  useWalletDetails,
  useWalletInfo,
} from '@cowprotocol/wallet'
import { useWeb3React } from '@web3-react/core'

import ReactGA from 'react-ga4'
import { useLocation } from 'react-router-dom'
import { getCLS, getFCP, getFID, getLCP, Metric } from 'web-vitals'

import { useInjectedWidgetMetaData } from 'modules/injectedWidget'

import { useGetMarketDimension } from './useGetMarketDimension'

export function sendTiming(timingCategory: any, timingVar: any, timingValue: any, timingLabel: any) {
  return googleAnalytics.gaCommandSendTiming(timingCategory, timingVar, timingValue, timingLabel)
}

function sendWebVitals({ name, delta, id }: Metric) {
  sendTiming('Web Vitals', name, Math.round(name === 'CLS' ? delta * 1000 : delta), id)
}

export function reportWebVitals() {
  getFCP(sendWebVitals)
  getFID(sendWebVitals)
  getLCP(sendWebVitals)
  getCLS(sendWebVitals)
}

export function initGATracker() {
  // typed as 'any' in react-ga4 -.-
  googleAnalytics.ga((tracker: any) => {
    if (!tracker) return

    const clientId = tracker.get('clientId')
    window.localStorage.setItem(GOOGLE_ANALYTICS_CLIENT_ID_STORAGE_KEY, clientId)
  })
}

let initiatedPixel = false

// tracks web vitals and pageviews
export function useAnalyticsReporter() {
  const { pathname, search } = useLocation()

  // Handle chain id custom dimension
  const { connector } = useWeb3React()
  const { chainId, account } = useWalletInfo()
  const { walletName: _walletName } = useWalletDetails()
  const injectedWidgetMetaData = useInjectedWidgetMetaData()
  const prevAccount = usePrevious(account)

  const marketDimension = useGetMarketDimension()

  useEffect(() => {
    // custom dimension 1 - chainId
    googleAnalytics.setDimension(Dimensions.chainId, chainId)
  }, [chainId])

  // Handle wallet name custom dimension
  const connection = getWeb3ReactConnection(connector)
  const isMetaMask = getIsMetaMask()

  const walletName = _walletName || getConnectionName(connection.type, isMetaMask)

  const injectedWidgetAppId = injectedWidgetMetaData?.appCode

  useEffect(() => {
    // Custom dimension 2 - walletname
    googleAnalytics.setDimension(Dimensions.walletName, account ? walletName : 'Not connected')

    // Custom dimension 4 - user id - because ReactGA.set might not be working
    const userId = account ? `"${account}"` : ''
    googleAnalytics.setDimension(Dimensions.userAddress, userId)
    ReactGA.set({ userId })

    // Handle pixel tracking on wallet connection
    if (!prevAccount && account) {
      sendAllPixels(PixelEvent.CONNECT_WALLET)
    }
  }, [account, walletName, prevAccount])

  useEffect(() => {
    // Custom dimension 5 - market
    googleAnalytics.setDimension(Dimensions.market, marketDimension)
  }, [marketDimension])

  useEffect(() => {
    // Custom dimension 6 - injected widget app id
    googleAnalytics.setDimension(Dimensions.injectedWidgetAppId, injectedWidgetAppId)
  }, [injectedWidgetAppId])

  useEffect(() => {
    googleAnalytics.pageview(`${pathname}${search}`)
  }, [pathname, search])

  useEffect(() => {
    reportWebVitals()
    initGATracker()
  }, [])

  // Handle initiate pixel tracking
  useEffect(() => {
    if (!initiatedPixel) {
      // // Init all pixels
      // const enablePixelFunctions = [
      //   enablePixelFacebook,
      //   enablePixelMicrosoft,
      //   enablePixelPaved,
      //   enablePixelReddit,
      //   enablePixelTwitter,
      // ]
      // enablePixelFunctions.forEach((enablePixel) => enablePixel())

      // Sent
      sendAllPixels(PixelEvent.INIT)

      initiatedPixel = true
    }
  }, [])
}
