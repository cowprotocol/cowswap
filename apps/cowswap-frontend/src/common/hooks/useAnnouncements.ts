import { useAtomValue } from 'jotai'
import { useMemo } from 'react'

import { isProdLike } from '@cowprotocol/common-utils'
import { Announcement, Announcements, announcementsAtom } from '@cowprotocol/core'
import { CowEnv, SupportedChainId } from '@cowprotocol/cow-sdk'

function getAnnouncementSpecificity(chainId: SupportedChainId, env: CowEnv, announcement: Announcement): number {
  let specificity = 0

  const matchesChain = announcement.chainIds.some((announcementChain) => announcementChain === chainId)
  const matchesEnv = announcement.envs.some((announcementEnv) => announcementEnv === env)
  const matchesEveryChain = announcement.chainIds.length === 0
  const matchesEveryEnv = announcement.envs.length === 0

  if (matchesChain) specificity += 2
  if (matchesEnv) specificity += 2
  if (matchesEveryChain) specificity += 1
  if (matchesEveryEnv) specificity += 1

  return specificity
}

function useAnnouncements(chainId: SupportedChainId): Announcements {
  const allAnnouncements = useAtomValue(announcementsAtom)

  return useMemo(() => {
    const env = isProdLike ? 'prod' : 'staging'

    const filtered = allAnnouncements
      .filter((announcement) => {
        const showForEveryChain = announcement.chainIds.length === 0
        const showForEveryEnv = announcement.envs.length === 0

        const matchesChainId = announcement.chainIds.some((announcementChain) => announcementChain === chainId)
        const matchesEnv = announcement.envs.some((announcementEnv) => announcementEnv === env)

        return (showForEveryChain || matchesChainId) && (showForEveryEnv || matchesEnv)
      })
      .sort((a, b) => {
        const specificityA = getAnnouncementSpecificity(chainId, env, a)
        const specificityB = getAnnouncementSpecificity(chainId, env, b)

        return specificityB - specificityA
      })

    return filtered
  }, [chainId, allAnnouncements])
}

export function useCriticalAnnouncements(chainId: SupportedChainId): Announcements {
  const announcements = useAnnouncements(chainId)

  return announcements.filter(({ isCritical }) => isCritical)
}

export function useNonCriticalAnnouncements(chainId: SupportedChainId): Announcements {
  const announcements = useAnnouncements(chainId)

  return announcements.filter(({ isCritical }) => !isCritical)
}
