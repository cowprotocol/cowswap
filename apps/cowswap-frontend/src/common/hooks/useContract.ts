import { useMemo } from 'react'

import {
  CoWSwapEthFlow,
  CoWSwapEthFlowAbi,
  Eip2612Abi,
  Erc1155,
  Erc1155Abi,
  Erc20,
  Erc20Abi,
  Erc20Bytes32Abi,
  Erc721,
  Erc721Abi,
  GPv2Settlement,
  GPv2SettlementAbi,
  UniswapInterfaceMulticall,
  UniswapInterfaceMulticallAbi,
  VCow,
  vCowAbi,
  Weth,
  WethAbi,
} from '@cowprotocol/abis'
import {
  COWSWAP_ETHFLOW_CONTRACT_ADDRESS,
  GP_SETTLEMENT_CONTRACT_ADDRESS,
  MULTICALL_ADDRESS,
  V_COW_CONTRACT_ADDRESS,
  WRAPPED_NATIVE_CURRENCIES,
} from '@cowprotocol/common-const'
import { getContract, isEns, isProd, isStaging } from '@cowprotocol/common-utils'
import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { useWalletInfo } from '@cowprotocol/wallet'
import { Contract } from '@ethersproject/contracts'
import type { JsonRpcProvider } from '@ethersproject/providers'
import { Web3Provider } from '@ethersproject/providers'
import { useWeb3React } from '@web3-react/core'

const { abi: MulticallABI } = UniswapInterfaceMulticallAbi

// returns null on errors
export function useContract<T extends Contract = Contract>(
  addressOrAddressMap: string | { [chainId: number]: string } | undefined,
  ABI: any,
  withSignerIfPossible = true,
  customProvider?: Web3Provider
): T | null {
  const { provider: defaultProvider } = useWeb3React()
  const { account, chainId } = useWalletInfo()
  const provider = customProvider || defaultProvider

  return useMemo(() => {
    if (!addressOrAddressMap || !ABI || !provider || !chainId) return null
    let address: string | undefined
    if (typeof addressOrAddressMap === 'string') address = addressOrAddressMap
    else address = addressOrAddressMap[chainId]
    if (!address) return null
    try {
      return getContract(address, ABI, provider, withSignerIfPossible && account ? account : undefined)
    } catch (error: any) {
      console.error('Failed to get contract', error)
      return null
    }
  }, [addressOrAddressMap, ABI, provider, chainId, withSignerIfPossible, account]) as T
}

export function useTokenContract(tokenAddress?: string, withSignerIfPossible?: boolean) {
  return useContract<Erc20>(tokenAddress, Erc20Abi, withSignerIfPossible)
}

export function useWETHContract(withSignerIfPossible?: boolean) {
  const { chainId } = useWalletInfo()
  return useContract<Weth>(
    chainId ? WRAPPED_NATIVE_CURRENCIES[chainId]?.address : undefined,
    WethAbi,
    withSignerIfPossible
  )
}

export function useERC721Contract(nftAddress?: string) {
  return useContract<Erc721>(nftAddress, Erc721Abi, false)
}

export function useERC1155Contract(nftAddress?: string) {
  return useContract<Erc1155>(nftAddress, Erc1155Abi, false)
}

export function useBytes32TokenContract(tokenAddress?: string, withSignerIfPossible?: boolean): Contract | null {
  return useContract(tokenAddress, Erc20Bytes32Abi, withSignerIfPossible)
}

export function useEIP2612Contract(tokenAddress?: string): Contract | null {
  return useContract(tokenAddress, Eip2612Abi, false)
}

export function useInterfaceMulticall(customProvider?: Web3Provider) {
  return useContract<UniswapInterfaceMulticall>(
    MULTICALL_ADDRESS,
    MulticallABI,
    false,
    customProvider
  ) as UniswapInterfaceMulticall
}

export function useEthFlowContract(): CoWSwapEthFlow | null {
  const { chainId } = useWalletInfo()

  const contractEnv = isProd || isStaging || isEns ? 'prod' : 'barn'

  const contractAddress = chainId ? COWSWAP_ETHFLOW_CONTRACT_ADDRESS[contractEnv][chainId] : undefined

  return useContract<CoWSwapEthFlow>(contractAddress, CoWSwapEthFlowAbi, true)
}

export function useGP2SettlementContract(): GPv2Settlement | null {
  const { chainId } = useWalletInfo()
  return useContract<GPv2Settlement>(
    chainId ? GP_SETTLEMENT_CONTRACT_ADDRESS[chainId] : undefined,
    GPv2SettlementAbi,
    true
  )
}

export function useVCowContract() {
  const { chainId } = useWalletInfo()
  return useContract<VCow>(chainId ? V_COW_CONTRACT_ADDRESS[chainId] : undefined, vCowAbi, true)
}

/**
 * Non-hook version of useContract
 */
function _getContract<T extends Contract = Contract>(
  addressOrAddressMap: string | { [chainId: number]: string } | undefined,
  ABI: any,
  withSignerIfPossible = true,
  provider?: JsonRpcProvider,
  account?: string,
  chainId?: SupportedChainId
): T | null {
  if (!addressOrAddressMap || !ABI || !provider || !chainId) return null
  let address: string | undefined
  if (typeof addressOrAddressMap === 'string') address = addressOrAddressMap
  else address = addressOrAddressMap[chainId]
  if (!address) return null
  try {
    return getContract(address, ABI, provider, withSignerIfPossible && account ? account : undefined) as T
  } catch (error: any) {
    console.error('Failed to get contract', error)
    return null
  }
}

/**
 * Non-hook version of useTokenContract
 */
export function getTokenContract(
  tokenAddress?: string,
  withSignerIfPossible?: boolean,
  provider?: JsonRpcProvider,
  account?: string,
  chainId?: SupportedChainId
): Erc20 | null {
  return _getContract<Erc20>(tokenAddress, Erc20Abi, withSignerIfPossible, provider, account, chainId)
}

/**
 * Non-hook version of useBytes32TokenContract
 */
export function getBytes32TokenContract(
  tokenAddress?: string,
  withSignerIfPossible?: boolean,
  provider?: JsonRpcProvider,
  account?: string,
  chainId?: SupportedChainId
): Contract | null {
  return _getContract(tokenAddress, Erc20Bytes32Abi, withSignerIfPossible, provider, account, chainId)
}
