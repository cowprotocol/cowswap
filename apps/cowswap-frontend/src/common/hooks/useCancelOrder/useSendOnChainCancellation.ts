import { useCallback } from 'react'

import { useWalletInfo } from '@cowprotocol/wallet'

import { useTransactionAdder } from 'legacy/state/enhancedTransactions/hooks'
import { Order } from 'legacy/state/orders/actions'
import { useRequestOrderCancellation, useSetOrderCancellationHash } from 'legacy/state/orders/hooks'

import { getIsEthFlowOrder } from 'modules/swap/containers/EthFlowStepper'
import { useSetPartOrderCancelling } from 'modules/twap/hooks/useSetPartOrderCancelling'

import { CancelledOrderInfo } from './onChainCancellation'
import { useGetOnChainCancellation } from './useGetOnChainCancellation'

export function useSendOnChainCancellation() {
  const { chainId } = useWalletInfo()
  const setOrderCancellationHash = useSetOrderCancellationHash()
  const cancelPendingOrder = useRequestOrderCancellation()
  const setPartOrderCancelling = useSetPartOrderCancelling()
  const addTransaction = useTransactionAdder()
  const getOnChainCancellation = useGetOnChainCancellation()

  const processCancelledOrder = useCallback(
    ({ txHash, orderId, sellTokenAddress, sellTokenSymbol }: CancelledOrderInfo) => {
      if (!chainId) return

      const isEthFlowOrder = getIsEthFlowOrder(sellTokenAddress)

      cancelPendingOrder({ id: orderId, chainId })
      setOrderCancellationHash({ chainId, id: orderId, hash: txHash })
      setPartOrderCancelling(orderId)

      if (isEthFlowOrder) {
        addTransaction({ hash: txHash, ethFlow: { orderId, subType: 'cancellation' } })
      } else {
        addTransaction({
          hash: txHash,
          onChainCancellation: { orderId, sellTokenSymbol: sellTokenSymbol || '' },
        })
      }
    },
    [chainId, cancelPendingOrder, setOrderCancellationHash, addTransaction, setPartOrderCancelling]
  )

  return useCallback(
    async (order: Order) => {
      const { sendTransaction } = await getOnChainCancellation(order)

      await sendTransaction(processCancelledOrder)
    },
    [processCancelledOrder, getOnChainCancellation]
  )
}
