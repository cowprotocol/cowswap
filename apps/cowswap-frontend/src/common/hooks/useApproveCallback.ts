import { useCallback } from 'react'

import { Erc20 } from '@cowprotocol/abis'
import { calculateGasMargin, getIsNativeToken } from '@cowprotocol/common-utils'
import { useWalletInfo } from '@cowprotocol/wallet'
import { BigNumber } from '@ethersproject/bignumber'
import { MaxUint256 } from '@ethersproject/constants'
import { TransactionResponse } from '@ethersproject/providers'
import { Currency, CurrencyAmount } from '@uniswap/sdk-core'

import { useTransactionAdder } from 'legacy/state/enhancedTransactions/hooks'

import { useTokenContract } from './useContract'

import { GAS_LIMIT_DEFAULT } from '../constants/common'

export async function estimateApprove(
  tokenContract: Erc20,
  spender: string,
  amountToApprove: CurrencyAmount<Currency>
): Promise<{
  approveAmount: BigNumber | string
  gasLimit: BigNumber
}> {
  try {
    return {
      approveAmount: MaxUint256,
      gasLimit: await tokenContract.estimateGas.approve(spender, MaxUint256),
    }
  } catch (e: any) {
    // Fallback: Attempt to set an approval for the maximum wallet balance (instead of the MaxUint256).
    // Some tokens revert if you try to use more than what you have.
    try {
      const approveAmount = amountToApprove.quotient.toString()

      return {
        approveAmount,
        gasLimit: await tokenContract.estimateGas.approve(spender, approveAmount),
      }
    } catch (error: any) {
      console.error(
        '[useApproveCallbackMod] Error estimating gas for approval. Using default gas limit ' +
          GAS_LIMIT_DEFAULT.toString(),
        error
      )

      return {
        approveAmount: MaxUint256,
        gasLimit: GAS_LIMIT_DEFAULT,
      }
    }
  }
}

export function useApproveCallback(
  amountToApprove?: CurrencyAmount<Currency>,
  spender?: string
): (summary?: string) => Promise<TransactionResponse | undefined> {
  const { chainId } = useWalletInfo()
  const currency = amountToApprove?.currency
  const token = currency && !getIsNativeToken(currency) ? currency : undefined
  const tokenContract = useTokenContract(token?.address)
  const addTransaction = useTransactionAdder()

  return useCallback(async () => {
    if (!chainId || !token || !tokenContract || !amountToApprove || !spender) {
      console.error('Wrong input for approve: ', { chainId, token, tokenContract, amountToApprove, spender })
      return
    }

    const estimation = await estimateApprove(tokenContract, spender, amountToApprove)
    return tokenContract
      .approve(spender, estimation.approveAmount, {
        gasLimit: calculateGasMargin(estimation.gasLimit),
      })
      .then((response: TransactionResponse) => {
        addTransaction({
          hash: response.hash,
          summary: amountToApprove.greaterThan('0') ? `Approve ${token.symbol}` : `Revoke ${token.symbol} approval`,
          approval: { tokenAddress: token.address, spender, amount: '0x' + amountToApprove.quotient.toString(16) },
        })
        return response
      })
  }, [chainId, token, tokenContract, amountToApprove, spender, addTransaction])
}
