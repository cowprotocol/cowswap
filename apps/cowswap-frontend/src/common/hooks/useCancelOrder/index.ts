import { useAtomValue, useSetAtom } from 'jotai'
import { useResetAtom } from 'jotai/utils'
import { useCallback } from 'react'

import { calculateGasMargin } from '@cowprotocol/common-utils'
import { Command } from '@cowprotocol/types'
import { useWalletDetails, useWalletInfo } from '@cowprotocol/wallet'

import { useCloseModal, useOpenModal } from 'legacy/state/application/hooks'
import { ApplicationModal } from 'legacy/state/application/reducer'
import { useGasPrices } from 'legacy/state/gas/hooks'
import { Order, OrderStatus } from 'legacy/state/orders/actions'

import { getIsEthFlowOrder } from 'modules/swap/containers/EthFlowStepper'
import { getSwapErrorMessage } from 'modules/trade/utils/swapErrorHelper'

import { useGetOnChainCancellation } from 'common/hooks/useCancelOrder/useGetOnChainCancellation'
import { computeOrderSummary } from 'common/updaters/orders/utils'
import { isOrderCancellable } from 'common/utils/isOrderCancellable'
import useNativeCurrency from 'lib/hooks/useNativeCurrency'

import { cancellationModalContextAtom, CancellationType, updateCancellationModalContextAtom } from './state'
import { useOffChainCancelOrder } from './useOffChainCancelOrder'
import { useSendOnChainCancellation } from './useSendOnChainCancellation'

export type UseCancelOrderReturn = Command | null

/**
 *
 * This hook encapsulates all the logic related to order cancellation, be it off or on-chain
 * It returns a callback which in turn returns a callback
 * It also handles the modal display logic
 *
 * The first callback takes an Order object to construct the callback which will trigger the cancellation modal display
 * It checks whether the order is eligible for cancellation and set's up which type of cancellation can be used (on or off-chain)
 * In case the order is not eligible, it returns null. This should be used to control whether a cancel button should be displayed
 */
export function useCancelOrder(): (order: Order) => UseCancelOrderReturn {
  const { chainId } = useWalletInfo()
  const { allowsOffchainSigning } = useWalletDetails()
  const openModal = useOpenModal(ApplicationModal.CANCELLATION)
  const closeModal = useCloseModal(ApplicationModal.CANCELLATION)
  const { isPendingSignature } = useAtomValue(cancellationModalContextAtom)
  const setContext = useSetAtom(updateCancellationModalContextAtom)
  const resetContext = useResetAtom(cancellationModalContextAtom)
  const offChainOrderCancel = useOffChainCancelOrder()
  const sendOnChainCancellation = useSendOnChainCancellation()
  const getOnChainTxInfo = useGetOnChainCancellation()
  const gasPrices = useGasPrices(chainId)
  const nativeCurrency = useNativeCurrency()

  return useCallback(
    (order: Order) => {
      // Check the 'cancellability'

      const isEthFlowOrder = getIsEthFlowOrder(order.inputToken.address)

      // 1. EthFlow orders will never be able to be cancelled offChain
      // 2. The wallet must support offChain singing
      // 3. The order must be PENDING
      const isOffChainCancellable = !isEthFlowOrder && allowsOffchainSigning && order?.status === OrderStatus.PENDING

      // When the order is not cancellable, there won't be a callback
      if (!isOrderCancellable(order)) {
        return null
      }

      // When dismissing the modal, close it and also reset context
      const onDismiss = () => {
        closeModal()
        resetContext()
      }

      // The callback to trigger the cancellation
      const triggerCancellation = async (type: CancellationType): Promise<void> => {
        const cancelFn = type === 'offChain' ? offChainOrderCancel : sendOnChainCancellation

        try {
          setContext({ isPendingSignature: true, error: null })
          // Actual cancellation is triggered here
          await cancelFn(order)
          onDismiss()
          // When done, dismiss the modal
        } catch (e: any) {
          onDismiss()
          if (!isPendingSignature) return

          const swapErrorMessage = getSwapErrorMessage(e?.body?.description || e)
          setContext({ error: swapErrorMessage })
        }
        setContext({ isPendingSignature: false })
      }

      // The callback returned that triggers the modal
      return () => {
        const summary = computeOrderSummary({
          orderFromStore: order,
          orderFromApi: null,
        })
        // Updates the cancellation context with details pertaining the order
        setContext({
          orderId: order.id,
          chainId,
          summary,
          defaultType: isOffChainCancellable ? 'offChain' : 'onChain',
          onDismiss,
          triggerCancellation,
          nativeCurrency,
        })
        // Display the actual modal
        openModal()
        // Estimate tx cost in case when OnChain cancellation is used
        getOnChainTxInfo(order).then(({ estimatedGas }) => {
          const gasPrice = +(gasPrices?.average || '0')
          const txCost = calculateGasMargin(estimatedGas).mul(gasPrice)

          setContext({ txCost })
        })
      }
    },
    [
      allowsOffchainSigning,
      chainId,
      closeModal,
      sendOnChainCancellation,
      offChainOrderCancel,
      openModal,
      resetContext,
      setContext,
      getOnChainTxInfo,
      gasPrices,
      nativeCurrency,
      isPendingSignature,
    ]
  )
}
