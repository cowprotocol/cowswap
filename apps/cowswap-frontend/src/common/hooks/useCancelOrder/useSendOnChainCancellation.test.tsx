import { PropsWithChildren } from 'react'

import { COW, NATIVE_CURRENCIES } from '@cowprotocol/common-const'
import { useWalletInfo } from '@cowprotocol/wallet'
import { BigNumber } from '@ethersproject/bignumber'

import { renderHook } from '@testing-library/react-hooks'

import { useTransactionAdder } from 'legacy/state/enhancedTransactions/hooks'
import { Order } from 'legacy/state/orders/actions'
import { useRequestOrderCancellation, useSetOrderCancellationHash } from 'legacy/state/orders/hooks'

import { useEthFlowContract, useGP2SettlementContract } from 'common/hooks/useContract'

import { useSendOnChainCancellation } from './useSendOnChainCancellation'

import { WithMockedWeb3 } from '../../../test-utils'

const chainId = 1
const settlementCancellationTxHash = '0xcfwj23g4fwe111'
const ethFlowCancellationTxHash = '0xcfwj23g4fwe222'

jest.mock('legacy/state/orders/hooks')
jest.mock('@cowprotocol/wallet', () => {
  return {
    ...jest.requireActual('@cowprotocol/wallet'),
    useWalletInfo: jest.fn().mockReturnValue({ chainId }),
  }
})
jest.mock('common/hooks/useContract', () => {
  return {
    ...jest.requireActual('common/hooks/useContract'),
    useEthFlowContract: jest.fn(),
    useGP2SettlementContract: jest.fn(),
  }
})
jest.mock('legacy/state/enhancedTransactions/hooks')
jest.mock('common/hooks/useAnalyticsReporter')

const orderMock = {
  id: 'xx1',
  buyToken: '0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB',
  receiver: '0x0000000000000000000000000000000000000000',
  sellAmount: '1',
  buyAmount: '2',
  appData: '0x001',
  validTo: 34245345432,
} as Order

const mockUseWalletInfo = useWalletInfo as jest.MockedFunction<typeof useWalletInfo>

const setOrderCancellationHash = jest.fn()
const mockUseSetOrderCancellationHash = useSetOrderCancellationHash as jest.MockedFunction<
  typeof useSetOrderCancellationHash
>
const requestOrderCancellation = jest.fn()
const mockUseRequestOrderCancellation = useRequestOrderCancellation as jest.MockedFunction<
  typeof useRequestOrderCancellation
>
const transactionAdder = jest.fn()
const mockUseTransactionAdder = useTransactionAdder as jest.MockedFunction<typeof useTransactionAdder>

const mockUseEthFlowContract = useEthFlowContract as jest.MockedFunction<typeof useEthFlowContract>
const mockUseGP2SettlementContract = useGP2SettlementContract as jest.MockedFunction<typeof useGP2SettlementContract>

const ethFlowInvalidationMock = jest.fn()
const settlementInvalidationMock = jest.fn()

const WithProviders = ({ children }: PropsWithChildren) => {
  return <WithMockedWeb3>{children}</WithMockedWeb3>
}

describe('useSendOnChainCancellation() + useGetOnChainCancellation()', () => {
  beforeEach(() => {
    jest.clearAllMocks()

    mockUseWalletInfo.mockReturnValue({ chainId })
    mockUseSetOrderCancellationHash.mockReturnValue(setOrderCancellationHash)
    mockUseRequestOrderCancellation.mockReturnValue(requestOrderCancellation)
    mockUseTransactionAdder.mockReturnValue(transactionAdder)

    ethFlowInvalidationMock.mockResolvedValue({ hash: ethFlowCancellationTxHash })

    mockUseEthFlowContract.mockReturnValue({
      estimateGas: {
        invalidateOrder: () => Promise.resolve(BigNumber.from(100)),
      },
      invalidateOrder: ethFlowInvalidationMock,
    } as any)

    settlementInvalidationMock.mockResolvedValue({ hash: settlementCancellationTxHash })

    mockUseGP2SettlementContract.mockReturnValue({
      estimateGas: {
        invalidateOrder: () => Promise.resolve(BigNumber.from(200)),
      },
      invalidateOrder: settlementInvalidationMock,
    } as any)
  })

  afterEach(() => {
    settlementInvalidationMock.mockClear()
    ethFlowInvalidationMock.mockClear()
  })

  it('When is ETH-flow order, then should call eth-flow contract', async () => {
    const { result } = renderHook(() => useSendOnChainCancellation(), { wrapper: WithProviders })

    await result.current({ ...orderMock, inputToken: NATIVE_CURRENCIES[chainId] })

    expect(ethFlowInvalidationMock).toHaveBeenCalledTimes(1)
    expect(ethFlowInvalidationMock.mock.calls[0]).toMatchSnapshot()
    expect(transactionAdder.mock.calls[0][0].hash).toBe(ethFlowCancellationTxHash)
  })

  it('When is NOT ETH-flow order, then should call settlement contract', async () => {
    const { result } = renderHook(() => useSendOnChainCancellation(), { wrapper: WithProviders })

    await result.current({ ...orderMock, inputToken: COW[chainId] })

    expect(settlementInvalidationMock).toHaveBeenCalledTimes(1)
    expect(settlementInvalidationMock.mock.calls[0]).toMatchSnapshot()
    expect(transactionAdder.mock.calls[0][0].hash).toBe(settlementCancellationTxHash)
  })

  describe('When a transaction is sent', () => {
    it('Then should change an order status, set a tx hash to order and add the transaction to store', async () => {
      const { result } = renderHook(() => useSendOnChainCancellation(), { wrapper: WithProviders })

      await result.current({ ...orderMock, inputToken: COW[chainId] })

      expect(transactionAdder).toHaveBeenCalledTimes(1)
      expect(transactionAdder.mock.calls[0]).toMatchSnapshot()
      expect(requestOrderCancellation).toHaveBeenCalledTimes(1)
      expect(requestOrderCancellation.mock.calls[0]).toMatchSnapshot()
      expect(setOrderCancellationHash).toHaveBeenCalledTimes(1)
      expect(setOrderCancellationHash.mock.calls[0]).toMatchSnapshot()
    })
  })
})
