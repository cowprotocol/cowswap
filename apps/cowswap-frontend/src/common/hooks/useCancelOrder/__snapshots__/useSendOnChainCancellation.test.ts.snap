// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSendOnChainCancellation() + useGetOnChainCancellation() When a transaction is sent Then should change an order status, set a tx hash to order and add the transaction to store 1`] = `
Array [
  Object {
    "hash": "0xcfwj23g4fwe111",
    "onChainCancellation": Object {
      "orderId": "xx1",
      "sellTokenSymbol": "COW",
    },
  },
]
`;

exports[`useSendOnChainCancellation() + useGetOnChainCancellation() When a transaction is sent Then should change an order status, set a tx hash to order and add the transaction to store 2`] = `
Array [
  Object {
    "chainId": 1,
    "id": "xx1",
  },
]
`;

exports[`useSendOnChainCancellation() + useGetOnChainCancellation() When a transaction is sent Then should change an order status, set a tx hash to order and add the transaction to store 3`] = `
Array [
  Object {
    "chainId": 1,
    "hash": "0xcfwj23g4fwe111",
    "id": "xx1",
  },
]
`;

exports[`useSendOnChainCancellation() + useGetOnChainCancellation() When is ETH-flow order, then should call eth-flow contract 1`] = `
Array [
  Object {
    "appData": "0x001",
    "buyAmount": "2",
    "buyToken": "0xDEf1CA1fb7FBcDC777520aa7f396b4E015F497aB",
    "feeAmount": undefined,
    "partiallyFillable": false,
    "quoteId": 0,
    "receiver": "0x0000000000000000000000000000000000000000",
    "sellAmount": "1",
    "validTo": "34245345432",
  },
  Object {
    "gasLimit": Object {
      "hex": "0x78",
      "type": "BigNumber",
    },
  },
]
`;

exports[`useSendOnChainCancellation() + useGetOnChainCancellation() When is NOT ETH-flow order, then should call settlement contract 1`] = `
Array [
  "xx1",
  Object {
    "gasLimit": Object {
      "hex": "0xf0",
      "type": "BigNumber",
    },
  },
]
`;
